cpp/include/raft/util/variant_utils.hpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <type_traits>
+#include <variant>
+
+namespace raft {
+
+template <typename variant1, typename variant2>
+struct concatenated_variant;
+
+template <typename... types1, typename... types2>
+struct concatenated_variant<std::variant<types1...>, std::variant<types2...>> {
+  using type = std::variant<types1..., types2...>;
+};
+
+template <typename variant1, typename variant2>
+using concatenated_variant_t = typename concatenated_variant<variant1, variant2>::type;
+
+template <typename visitor_t, typename variant_t, std::size_t index = std::size_t{}>
```

Comment:
What is the optimization here? Being able to provide the starting point? For example, providing `2` if we only want `managed` or `pinned` views

Reply:
That's an interesting thought, and I had not even thought about that potential use case (though I imagine it might get confusing without some additional wrapping). Cool! But not what I was going for. ;)

No, this is to deal with a limit on inlining that is accidentally imposed by the computational complexity requirements in the standard for `std::visit`. The short version is that the requirement essentially forces stdlib implementations to create a jump table when they could inline instead.

This implementation converts the dispatch to a bunch of if/else branches which _may_ still permit inlining. That being said, I actually do not currently use this anywhere in the current implementation, so I'm happy to just remove it from the PR.

Later on, I'd like to review each place that we use `std::visit` in the actual mdbuffer implementation and see if they could benefit from this optimization. I suspect for most of them, they just wouldn't benefit from inlining anyway, but it would be good to benchmark and make sure that the core mdbuffer implementation is as close to speed-of-light as possible.

Reply:
Aha! Now I'm understanding a bit better. There are actually a couple different issues that are getting pulled together here. Let's start with this one:

> For example, pinned should be able to provide an mdspan for all APIs that accept host/device/pinned.

There's actually a bug in the current implementation because [this line here](https://github.com/rapidsai/raft/blob/branch-23.12/cpp/include/raft/core/memory_type.hpp#L23) does not label pinned memory as device-accessible (Which blames to me. Oops!). I'll fix that up, but the point still stands for both pinned and managed memory.

Let's imagine you have three functions: `foo_host`, `foo_device`, and `foo_universal`. You want to invoke those functions with four different mdspans: `host_md`, `device_md`, `managed_md`, and `pinned_md`. The following wrappers will work for all 12 combinations:

```
template <typename mdspan_type>
void wrapper_host(mdspan_type input) {
   auto buf = mdbuffer(res, input, memory_type::host); // No copy for host/managed/pinned input; copy for device input
   foo_host(buf.view<memory_type::host>());
}

template <typename mdspan_type>
void wrapper_device(mdspan_type input) {
   auto buf = mdbuffer(res, input, memory_type::device); // No copy for device/managed/pinned input; copy for host input
   foo_device(buf.view<memory_type::device>());
}

template <typename mdspan_type>
void wrapper_universal(mdspan_type input) {
   auto buf = mdbuffer(res, input, memory_type::managed); // No copy for managed/pinned input; copy for host/device input
   foo_universal(buf.view<memory_type::managed>());
}
```
Those wrappers ensure that each of the foo functions can accept _any_ memory type, but let's take a look at how we would do runtime dispatching to whichever wrapper avoids copies. We'll use `std::visit` here first and then look at a custom dispatcher.

```
void wrapper(mdbuffer input) {
  std::visit([](auto view) {
    if constexpr (typename decltype(view)::accessor_type::is_host_device_accessible) {
      foo_universal(input.view<memory_type::managed>);
    } else if constexpr (typename decltype(view)::accessor_type::is_device_accessible) {
      foo_device(input.view<memory_type::device>());
    } else {
      foo_host(input.view<memory_type::host>());
    }
  }, data.view());
}
```
Or more practically speaking, since we rarely specialize these kinds of function calls for host/device accessibility but instead just dispatch to device methods for managed/pinned memory:
```
void wrapper(mdbuffer input) {
  std::visit([](auto view) {
    if constexpr (typename decltype(view)::accessor_type::is_device_accessible) {
      foo_device(input.view<memory_type::device>()); // invoked for device/managed/pinned input
    } else {
      foo_host(input.view<memory_type::host>()); // invoked for host input
    }
  }, data.view());
}
```
So really, we want to set up something that optimally avoids this boilerplate for the common task of dispatching based on where data are accessible from. I started to write out something based on cuDF's type dispatcher here in the comments, but it's complex enough, I'm going to post it separately and see what you think of the whole thing.

Reply:
Resolving to shift conversation to the dispatcher implementation.